Properly Serialize NULL Instant's

## Overview

`NULL` has to be carefully treated as the absence of values within the `CommitVerification` system is intentionally used with meaning.

## Root Cause

The error output from the issues.

### Original Issue

A student posted this image which was likely caused by this issue:

![image](https://github.com/user-attachments/assets/b442936f-f0b6-42a4-aea4-ebd6f838fa1f)

Backend stack traces:

```
edu.byu.cs.dataAccess.DataAccessException: Error inserting item into table submission
        at edu.byu.cs.dataAccess.sql.helpers.SqlReader.insertItem(SqlReader.java:128)
        at edu.byu.cs.dataAccess.sql.SubmissionSqlDao.insertSubmission(SubmissionSqlDao.java:89)
        at edu.byu.cs.autograder.Grader.run(Grader.java:106)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: edu.byu.cs.util.Serializer$SerializationException: java.lang.NullPointerException: Cannot invoke "java.time.Instant.toString()" because "instant" is null
        at edu.byu.cs.util.Serializer.serialize(Serializer.java:30)
        at edu.byu.cs.model.Submission.serializeObject(Submission.java:183)
        at edu.byu.cs.model.Submission.serializeCommitResult(Submission.java:179)
        at edu.byu.cs.dataAccess.sql.helpers.SqlReader.setValue(SqlReader.java:133)
        at edu.byu.cs.dataAccess.sql.helpers.SqlReader.insertItem(SqlReader.java:123)
        ... 7 common frames omitted
Caused by: java.lang.NullPointerException: Cannot invoke "java.time.Instant.toString()" because "instant" is null
        at edu.byu.cs.util.Serializer$InstantAdapter.write(Serializer.java:124)
        at edu.byu.cs.util.Serializer$InstantAdapter.write(Serializer.java:120)
        at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:70)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:196)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:368)
        at com.google.gson.Gson.toJson(Gson.java:842)
        at com.google.gson.Gson.toJson(Gson.java:812)
        at com.google.gson.Gson.toJson(Gson.java:759)
        at com.google.gson.Gson.toJson(Gson.java:736)
        at edu.byu.cs.util.Serializer.serialize(Serializer.java:28)
        ... 11 common frames omitted
```

_Originally posted by @19mdavenport in https://github.com/softwareconstruction240/autograder/pull/556#issuecomment-2691831715_

### Cascading Issue

Additionally, this code caused another error by assuming that `commitResult()` would never be NULL.

![image](https://github.com/user-attachments/assets/bf4afe68-780b-494d-8605-aee2e40d3924)

In this case, the error is especially painful because it also causes an error when attempting to manually approve the submission:

![image](https://github.com/user-attachments/assets/15be9542-c06b-4f8a-aa7b-95bcf55d3581)

_Originally posted by @frozenfrank in https://github.com/softwareconstruction240/autograder/issues/556#issuecomment-2692363139_

--- a/src/main/java/edu/byu/cs/autograder/git/CommitVerificationResult.java
+++ b/src/main/java/edu/byu/cs/autograder/git/CommitVerificationResult.java
@@ -39,8 +39,8 @@ public record CommitVerificationResult(
         @Nullable Collection<String> warningMessages,
 
         // Debug info
-        @NonNull Instant minAllowedThreshold,
-        @NonNull Instant maxAllowedThreshold,
+        Instant minAllowedThreshold,
+        Instant maxAllowedThreshold,
         @NonNull String headHash,
         String tailHash
 ) {

--- a/src/main/java/edu/byu/cs/autograder/git/CommitVerificationResult.java
+++ b/src/main/java/edu/byu/cs/autograder/git/CommitVerificationResult.java
@@ -12,6 +12,9 @@ import java.util.Collection;
  * and some debug information from the Commit Verification System.
  *
  * @see edu.byu.cs.model.Submission.ScoreVerification#verifiedStatus() for more information about `penaltyPct`.
+ * <br>
+ * Many of these fields are clearly and intentionally labelled as {@link Nullable} or {@link NonNull}.
+ * All the other fields are primitive types and can never be set to null.
  *
  * @param verified Officially reports the verified status of the submission.
  * @param isCachedResponse Whether this was freshly generated, or passed through from a previous decision.
@@ -22,10 +25,10 @@ import java.util.Collection;
  * @param penaltyPct A reduction to the phase, if any, when a TA approved an unapproved score. [0-100]
  * @param failureMessage A string that will be presented to the use when this the result is not verified.
  * @param warningMessages A nullable, potentially empty collection of warning messages that should be presented to the user.
- * @param minAllowedThreshold Debug purposes. The min timestamp considered.
- * @param maxAllowedThreshold Debug. The maximum timestamp considered.
- * @param headHash Debug. The head hash evaluated. Not null.
- * @param tailHash Debug. The tail hash evaluated. Sometimes null.
+ * @param minAllowedThreshold Debug purposes. The min timestamp considered. Null when the evaluation is not performed, such as when <pre>isCachedResponse</pre> is true.
+ * @param maxAllowedThreshold Debug. The maximum timestamp considered. Similar conditions to above.
+ * @param headHash Debug. The head hash evaluated. Not null. Even when evaluation is skipped, this still represents the hash at the time of consideration.
+ * @param tailHash Debug. The tail hash evaluated. Sometimes null if there is no previous passing submission, or if no evaluation was completed.
  */
 public record CommitVerificationResult(
         @NonNull boolean verified,
@@ -39,10 +42,10 @@ public record CommitVerificationResult(
         @Nullable Collection<String> warningMessages,
 
         // Debug info
-        Instant minAllowedThreshold,
-        Instant maxAllowedThreshold,
+        @Nullable Instant minAllowedThreshold,
+        @Nullable Instant maxAllowedThreshold,
         @NonNull String headHash,
-        String tailHash
+        @Nullable String tailHash
 ) {
     public CommitVerificationReport toReport(CommitVerificationContext context) {
         return new CommitVerificationReport(context, this);

--- a/src/main/java/edu/byu/cs/autograder/git/GitHelper.java
+++ b/src/main/java/edu/byu/cs/autograder/git/GitHelper.java
@@ -215,8 +215,7 @@ public class GitHelper {
         return new CommitVerificationResult(
                 verified, true,
                 0, 0, 0, false, originalPenaltyPct, failureMessage, null,
-                firstPassingSubmission.commitResult().minAllowedThreshold(), firstPassingSubmission.commitResult().maxAllowedThreshold(),
-                headHash, null
+                null, null, headHash, null
         ).toReport(null);
     }
 

--- a/src/main/java/edu/byu/cs/util/Serializer.java
+++ b/src/main/java/edu/byu/cs/util/Serializer.java
@@ -6,6 +6,7 @@ import com.google.gson.stream.JsonToken;
 import com.google.gson.stream.JsonWriter;
 import edu.byu.cs.canvas.model.CanvasRubricAssessment;
 import edu.byu.cs.canvas.model.CanvasRubricItem;
+import org.eclipse.jgit.annotations.Nullable;
 
 import java.io.IOException;
 import java.io.Reader;
@@ -120,12 +121,16 @@ public class Serializer {
     private static class InstantAdapter extends TypeAdapter<Instant> {
 
         @Override
-        public void write(JsonWriter jsonWriter, Instant instant) throws IOException {
-            jsonWriter.value(instant.toString());
+        public void write(JsonWriter jsonWriter, @Nullable Instant instant) throws IOException {
+            jsonWriter.value(instant == null ? null : instant.toString());
         }
 
         @Override
         public Instant read(JsonReader jsonReader) throws IOException {
+            if(jsonReader.peek() == JsonToken.NULL) {
+                jsonReader.nextNull();
+                return null;
+            }
             return Instant.parse(jsonReader.nextString());
         }
     }

--- a/src/main/java/edu/byu/cs/util/Serializer.java
+++ b/src/main/java/edu/byu/cs/util/Serializer.java
@@ -6,7 +6,7 @@ import com.google.gson.stream.JsonToken;
 import com.google.gson.stream.JsonWriter;
 import edu.byu.cs.canvas.model.CanvasRubricAssessment;
 import edu.byu.cs.canvas.model.CanvasRubricItem;
-import org.eclipse.jgit.annotations.Nullable;
+import org.jetbrains.annotations.NotNull;
 
 import java.io.IOException;
 import java.io.Reader;
@@ -78,22 +78,6 @@ public class Serializer {
         }
     }
 
-    private static class ZonedDateTimeAdapter extends TypeAdapter<ZonedDateTime> {
-        @Override
-        public void write(JsonWriter jsonWriter, ZonedDateTime zonedDateTime) {
-        }
-
-        @Override
-        public ZonedDateTime read(JsonReader jsonReader) throws IOException {
-            if (jsonReader.peek() == JsonToken.NULL) {
-                jsonReader.nextNull();
-                return null;
-            }
-            ZonedDateTime utc = ZonedDateTime.parse(jsonReader.nextString());
-            // TODO: Read timezone from dynamic location
-            return utc.withZoneSameInstant(ZoneId.of("America/Denver"));
-        }
-    }
 
     private static class RubricAssessmentAdapter implements JsonDeserializer<CanvasRubricAssessment> {
         @Override
@@ -118,20 +102,41 @@ public class Serializer {
         }
     }
 
-    private static class InstantAdapter extends TypeAdapter<Instant> {
-
+    private abstract static class NullSafeTypeAdapter<T> extends TypeAdapter<T> {
         @Override
-        public void write(JsonWriter jsonWriter, @Nullable Instant instant) throws IOException {
-            jsonWriter.value(instant == null ? null : instant.toString());
+        public void write(JsonWriter jsonWriter, T t) throws IOException {
+            jsonWriter.value(t == null ? null : writeNotNull(t));
+        }
+
+        private String writeNotNull(@NotNull T t) {
+            return t.toString();
         }
 
         @Override
-        public Instant read(JsonReader jsonReader) throws IOException {
-            if(jsonReader.peek() == JsonToken.NULL) {
+        public T read(JsonReader jsonReader) throws IOException {
+            if (jsonReader.peek() == JsonToken.NULL) {
                 jsonReader.nextNull();
                 return null;
             }
+            return readNotNull(jsonReader);
+        }
+
+        protected abstract T readNotNull(JsonReader jsonReader) throws IOException;
+    }
+
+    private static class InstantAdapter extends NullSafeTypeAdapter<Instant> {
+        @Override
+        protected Instant readNotNull(JsonReader jsonReader) throws IOException {
             return Instant.parse(jsonReader.nextString());
+        }
+    }
+
+    private static class ZonedDateTimeAdapter extends NullSafeTypeAdapter<ZonedDateTime> {
+        @Override
+        protected ZonedDateTime readNotNull(JsonReader jsonReader) throws IOException {
+            ZonedDateTime utc = ZonedDateTime.parse(jsonReader.nextString());
+            // TODO: Read timezone from dynamic location
+            return utc.withZoneSameInstant(ZoneId.of("America/Denver"));
         }
     }
 }

